security/advancedtls/advancedtls_integration_test.go

```diff
@@ -0,0 +1,509 @@
+	serverTrust1 *x509.CertPool
+	serverTrust2 *x509.CertPool
+}
+
+// loadCerts function is used to load test certificates at the beginning of each integration test.
+func (cs *certStore) loadCerts() error {
+	var err error
+	cs.clientPeer1, err = tls.LoadX509KeyPair(testdata.Path("client_cert_1.pem"),
+		testdata.Path("client_key_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientPeer2, err = tls.LoadX509KeyPair(testdata.Path("client_cert_2.pem"),
+		testdata.Path("client_key_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverPeer1, err = tls.LoadX509KeyPair(testdata.Path("server_cert_1.pem"),
+		testdata.Path("server_key_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverPeer2, err = tls.LoadX509KeyPair(testdata.Path("server_cert_2.pem"),
+		testdata.Path("server_key_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientTrust1, err = readTrustCert(testdata.Path("client_trust_cert_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.clientTrust2, err = readTrustCert(testdata.Path("client_trust_cert_2.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverTrust1, err = readTrustCert(testdata.Path("server_trust_cert_1.pem"))
+	if err != nil {
+		return err
+	}
+	cs.serverTrust2, err = readTrustCert(testdata.Path("server_trust_cert_2.pem"))
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// serverImpl is used to implement pb.GreeterServer.
+type serverImpl struct{}
+
+// SayHello is a simple implementation of pb.GreeterServer.
+func (s *serverImpl) SayHello(ctx context.Context, in *pb.HelloRequest) (*pb.HelloReply, error) {
+	return &pb.HelloReply{Message: "Hello " + in.Name}, nil
+}
+
+func callAndVerifyInSameRPC(msg string, client pb.GreeterClient, shouldFail bool) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+	_, err := client.SayHello(ctx, &pb.HelloRequest{Name: msg})
+	if want, got := shouldFail == true, err != nil; got != want {
+		return fmt.Errorf("want and got mismatch,  want shouldFail=%v, got fail=%v, rpc error: %v", want, got, err)
+	}
+	return nil
+}
+
+func callAndVerifyWithNewRPC(connCtx context.Context, msg string, creds credentials.TransportCredentials, shouldFail bool) (*grpc.ClientConn, pb.GreeterClient, error) {
+	var conn *grpc.ClientConn
+	var err error
+	// If we want the test to fail, we establish a non-blocking connection to avoid it hangs and killed by the context.
+	if shouldFail {
+		conn, err = grpc.DialContext(connCtx, address, grpc.WithTransportCredentials(creds))
+		if err != nil {
+			return nil, nil, fmt.Errorf("client failed to connect to %s. Error: %v", address, err)
+		}
+	} else {
+		conn, err = grpc.DialContext(connCtx, address, grpc.WithTransportCredentials(creds), grpc.WithBlock())
+		if err != nil {
+			return nil, nil, fmt.Errorf("client failed to connect to %s. Error: %v", address, err)
+		}
+	}
+	greetClient := pb.NewGreeterClient(conn)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```

Comment:
It looks like we can call `callAndVerifyInSameRPC` here since it's the same code.